c2corg_ui/static/js/searchdirective.js

```diff
@@ -0,0 +1,184 @@
+    highlight: true,
+    hint: true,
+    minLength: 2
+  });
+
+  /**
+   * @type {Array.<TypeaheadDataset>}
+   * @export
+   */
+  this.datasets = [
+    this.createDataset_('waypoints', 'Waypoints'),
+    this.createDataset_('routes', 'Routes')
+  ];
+
+  /**
+   * @type {ngeox.SearchDirectiveListeners}
+   * @export
+   */
+  this.listeners = /** @type {ngeox.SearchDirectiveListeners} */ ({
+    select: angular.bind(this, app.SearchController.select_)
+  });
+};
+
+
+/**
+ * @param {string} type The document type.
+ * @param {string} title The category title.
+ * @return {TypeaheadDataset} A data set.
+ * @private
+ */
+app.SearchController.prototype.createDataset_ = function(type, title) {
+  var bloodhoundEngine = this.createAndInitBloodhound_(type);
+  return /** @type {TypeaheadDataset} */({
+    source: bloodhoundEngine.ttAdapter(),
+    display: function(suggestion) {
+      return suggestion['label'];
+    },
+    limit: 20,
+    templates: {
+      header: angular.bind(this, function() {
+        return '<div class="header">' +
+            this.gettextCatalog_.getString(title) + '</div>';
+      }),
+      suggestion: function(suggestion) {
+        return '<p>' + suggestion['label'] + '</p>';
+      }
+      // Note: The templates for `notFound` and `pending` are not set, because
+      // they would be shown for each dataset.
+    }
+  });
+};
+
+
+/**
+ * @param {string} type The document type.
+ * @return {Bloodhound} The bloodhound engine.
+ * @private
+ */
+app.SearchController.prototype.createAndInitBloodhound_ = function(type) {
+  var url = this.apiUrl_ + '/search?q=%QUERY';
+  var bloodhound = new Bloodhound(/** @type {BloodhoundOptions} */({
+    limit: 10,
+    queryTokenizer: Bloodhound.tokenizers.whitespace,
+    datumTokenizer: Bloodhound.tokenizers.obj.whitespace('label'),
+    remote: {
+      url: url,
+      wildcard: '%QUERY',
+      rateLimitWait: 50,
+      filter: angular.bind(this, function(resp) {
+        var documents = resp[type]['documents'];
+        var currentCulture = this.gettextCatalog_.currentLanguage;
+        var cultures = this.cultures_;
+        return documents.map(function(document) {
+          var locale = app.utils.getBestLocale(
+              document, currentCulture, cultures);
+          document['label'] = locale['title'];
+          if (currentCulture !== locale['culture']) {
+            document['label'] += ' (' + locale['culture'] + ')';
+          }
+          document['bestCulture'] = locale['culture'];
+          document['type'] = type;
+          return document;
+        });
+      })
+    }
+  }));
+  bloodhound.initialize();
+  return bloodhound;
+};
+
+
+/**
+ * @param {jQuery.Event} event Event.
+ * @param {Object} document Suggested document.
+ * @param {TypeaheadDataset} dataset Dataset.
+ * @this {app.SearchController}
+ * @private
+ */
+app.SearchController.select_ = function(event, document, dataset) {
+  var culture = document['bestCulture'];
```

Comment:
Minor suggestion: maybe use `doc` to avoid shadowing the global `document` object.